sim/midas/src/main/scala/midas/models/dram/IngressUnit.scala

```diff
@@ -94,28 +106,60 @@ class IngressModule(val cfg: BaseConfig)(implicit val p: Parameters) extends Mod
                        ((awCredits.value < wCredits.value) && awCredits.inc) ||
                         awCredits.inc && wCredits.inc
 
+  val read_req_done = arQueue.io.enq.fire()
   when (!io.relaxed) {
     Seq(awCredits, wCredits) foreach { _.dec := write_req_done }
   }
 
+  val readMatcher = Module(new FIFOAddressMatcher(cfg.maxReads*2, addrWidth)).io
+  readMatcher.enq.valid := !ingressUnitStall && io.nastiInputs.hBits.ar.valid
+  readMatcher.enq.bits := io.nastiInputs.hBits.ar.bits.addr
+  readMatcher.match_address := io.nastiInputs.hBits.aw.bits.addr
 
-  val read_req_done = arQueue.io.enq.fire()
+  val writeMatcher = Module(new FIFOAddressMatcher(cfg.maxWrites*2, addrWidth)).io
+  writeMatcher.enq.valid := !ingressUnitStall && io.nastiInputs.hBits.aw.valid
+  writeMatcher.enq.bits := io.nastiInputs.hBits.aw.bits.addr
+  writeMatcher.match_address := io.nastiInputs.hBits.aw.bits.addr
 
+  arIDQueue.io.enq.valid := tFireHelper.fire(arIDQueue.io.enq.ready) && io.nastiInputs.hBits.ar.valid
+  arIDQueue.io.enq.bits := io.nastiInputs.hBits.ar.bits.id
+  awIDQueue.io.enq.valid := tFireHelper.fire(awIDQueue.io.enq.ready) && io.nastiInputs.hBits.aw.valid
+  awIDQueue.io.enq.bits := io.nastiInputs.hBits.aw.bits.id
   // FIFO that tracks the relative order of reads and writes are they are received 
   // bit 0 = Read, bit 1 = Write
   val xaction_order = Module(new DualQueue(Bool(), cfg.maxReads + cfg.maxWrites))
-  xaction_order.io.enqA.valid := read_req_done
+  val xaction_order_collisions = Module(new DualQueue(Bool(), cfg.maxReads + cfg.maxWrites))
+  xaction_order.io.enqA.valid := read_req_done && !collisions
   xaction_order.io.enqA.bits := true.B
-  xaction_order.io.enqB.valid := write_req_done
+  xaction_order.io.enqB.valid := write_req_done && !collisions
   xaction_order.io.enqB.bits := false.B
+  xaction_order_collisions.io.enqA.valid := read_req_done
+  xaction_order_collisions.io.enqA.bits := writeMatcher.hit
+  xaction_order_collisions.io.enqB.valid := write_req_done
+  xaction_order_collisions.io.enqB.bits := readMatcher.hit
+
+  val writeCollisions = SatUpDownCounter(cfg.maxWrites)
 
-  val do_hread = io.relaxed ||
+  val readCollisions = SatUpDownCounter(cfg.maxReads)
+
+  writeCollisions.inc := awQueue.io.enq.fire() && readMatcher.hit && io.nastiInputs.hBits.aw.valid 
+  writeCollisions.dec := xaction_order.io.deq.valid && xaction_order.io.deq.bits &&
+                           xaction_order_collisions.io.deq.valid && xaction_order_collisions.io.deq.bits
+  readCollisions.inc := arQueue.io.enq.fire() && writeMatcher.hit && io.nastiInputs.hBits.ar.valid
+  readCollisions.dec := !xaction_order.io.deq.valid && xaction_order.io.deq.bits &&
+                           xaction_order_collisions.io.deq.valid && xaction_order_collisions.io.deq.bits
+
+  val collisions = writeCollisions.value === 0.U && readCollisions.value === 0.U
+
+
+  val do_hread = io.relaxed || collisions ||
```

Comment:
I can't remember what i said, but i think i'd prefer that the existing conservative behavior also exist. In other words, the "relaxed" version would be yours. in that case the logic here should be changed (io.relaxed && collisions) || <...>

We can add a "fully relaxed mode" if we wanted, but i don't think it would see use. 
This way if there's a bug in the scheduler we can fall back on the legacy behavior, which we know works.